crates/bindings-csharp/BSATN.Codegen/Diag.cs

```diff
@@ -0,0 +1,232 @@
+    )> EnumWithExplicitValues =
+        new(
+            group,
+            "[SpacetimeDB.Type] enums cannot have explicit values",
+            ctx =>
+                $"{ctx.@enum.Identifier}.{ctx.enumMember.Identifier} has an explicit value {ctx.equalsValue.Value} which is not allowed in SpacetimeDB enums.",
+            ctx => ctx.equalsValue
+        );
+
+    public static readonly ErrorDescriptor<EnumDeclarationSyntax> EnumTooManyVariants =
+        new(
+            group,
+            "[SpacetimeDB.Type] enums are limited to 256 variants",
+            @enum =>
+                $"{@enum.Identifier} has {@enum.Members.Count} variants which is more than the allowed 256 variants for SpacetimeDB enums.",
+            @enum => @enum.Members[256]
+        );
+
+    public static readonly ErrorDescriptor<INamedTypeSymbol> TaggedEnumInlineTuple =
+        new(
+            group,
+            "Tagged enum variants must be declared with inline tuples",
+            baseType =>
+                $"{baseType} does not have the expected format SpacetimeDB.TaggedEnum<(TVariant1 v1, ..., TVariantN vN)>.",
+            baseType => baseType
+        );
+
+    public static readonly ErrorDescriptor<IFieldSymbol> TaggedEnumField =
+        new(
+            group,
+            "Tagged enums cannot have instance fields",
+            field =>
+                $"{field.Name} is an instance field, which are not permitted inside SpacetimeDB tagged enums.",
+            field => field
+        );
+
+    public static readonly ErrorDescriptor<TypeParameterListSyntax> TypeParams =
+        new(
+            group,
+            "Type parameters are not yet supported",
+            typeParams => $"Type parameters {typeParams} are not supported in SpacetimeDB types.",
+            typeParams => typeParams
+        );
+}
+
+// This class is used to collect diagnostics during parsing and return them as a combined result.
+//
+// It's necessary because Roslyn doesn't let incremental generators emit diagnostics while parsing,
+// only while emitting - which means we need to collect them during parsing and store in some data
+// structure (this one) and report them later, when we have the emitter context.
+//
+// Diagnostics are not kept inside the parsed data structure to make the main data cacheable even
+// when the diagnostic locations and details change between reruns.
+public record ParseResult<T>(T? Parsed, EquatableArray<Diagnostic> Diag)
+    where T : IEquatable<T>;
+
+public class DiagReporter
+{
+    private readonly ImmutableArray<Diagnostic>.Builder builder =
+        ImmutableArray.CreateBuilder<Diagnostic>();
+
+    public void Report<TContext>(ErrorDescriptor<TContext> descriptor, TContext ctx)
+    {
+        builder.Add(descriptor.ToDiag(ctx));
+    }
+
+    private DiagReporter() { }
+
+    private static readonly ErrorDescriptor<(SyntaxNode node, Exception e)> InternalError =
+        new(
+            new("STDBINT", "SpacetimeDB.Internal"),
```

Comment:
STDBINT? What about "STDB._" ?

Reply:
And can these be lowercase? I don't think diagnostics should scream at users :)

Reply:
These are diagnostic IDs, they're uppercase letters + 4 digits by convention, and most tooling likely expects them to remain so. The 2nd one is the "human-readable" category.